include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+        uint32_t _residentStorageCount,
+        uint32_t _pgTabSzxy_log2 = 8u,
+        uint32_t _pgTabLayers = 256u,
+        uint32_t _pgSzxy_log2 = 7u,
+        uint32_t _tilePadding = 9u,
+        uint32_t _maxAllocatableTexSz_log2 = 14u
+    ) : IVirtualTexture(
+        _pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _tilePadding
+    ) {
+        m_pageTable = createPageTable(_pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
+        initResidentStorage(_residentStorageParams, _residentStorageCount);
+    }
+
+    STextureData pack(const ICPUImage* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        uint32_t smplrIndex = 0u;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return STextureData::invalid();
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+
+            SamplerArray* views = nullptr;
+            if (isFloatingPointFormat(format)||isNormalizedFormat(format)||isScaledFormat(format))
+                views = &m_fsamplers;
+            else if (isSignedFormat(format))
+                views = &m_isamplers;
+            else
+                views = &m_usamplers;
+            auto view_it = std::find_if(views->views->begin(), views->views->end(), [format](const core::smart_refctd_ptr<ICPUImageView>& _view) {return _view->getCreationParameters().format==format;});
+            if (view_it==views->views->end()) //no physical page texture view/sampler for requested format
+                return STextureData::invalid();
+            smplrIndex = std::distance(views->views->begin(), view_it);
+        }
+        auto assignedLayers = storage->getPageTableLayers();
+
+        page_tab_offset_t pgtOffset = page_tab_offset_invalid();
+        for (auto it = assignedLayers.first; it != assignedLayers.second; ++it)
+        {
+            pgtOffset = alloc(_img, _subres, *it);
+            if ((pgtOffset==page_tab_offset_invalid()).all())
+                continue;
+        }
+        if ((pgtOffset==page_tab_offset_invalid()).all())
+        {
+            if (m_freePageTableLayerIDs.empty())
+                return STextureData::invalid();
+            const uint32_t pgtLayer = m_freePageTableLayerIDs.top();
+            m_freePageTableLayerIDs.pop();
+            pgtOffset = alloc(_img, _subres, pgtLayer);
+            if ((pgtOffset==page_tab_offset_invalid()).all())//this would be super weird but let's check
+                return STextureData::invalid();
+            storage->addPageTableLayer(pgtLayer);
+
+            (*m_layerToViewIndexMapping)[pgtLayer] = smplrIndex;
```

Comment:
https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files#r417924213

weren't you supposed to use this here?